blog/_src/posts/2018-12-02-java-transient.md

```diff
@@ -0,0 +1,361 @@
+```
+
+If generic-type erasure results in a type mismatch (e.g., in `sBox.get().charAt(0)` above),
+ the compiler inserts a cast.
+The inserted casts led to the runtime error in the previous example, and
+ provide the following run-time guarantee:
+
+- if `e` is an expression with static type `T` that evaluates to a value `v`,
+- then `v` is guaranteed to match the (bytecode) type `G(T)`
+
+
+## Discussion
+
+TypeScript, Reticulated Python, and Java 1.5.0 each improved the type system
+ of an existing language, but maintained backwards compatibility with existing
+ code.
+The name [migratory typing](http://drops.dagstuhl.de/opus/volltexte/2017/7120/)
+ describes this kind of language extension.
+
+> [Gradual typing](http://drops.dagstuhl.de/opus/volltexte/2015/5031/)
+> is a similar; a gradual type system starts with a statically-typed language
+> and adds dynamic typing in a principled way ([example](https://pleiad.cl/papers/2016/garciaAl-popl2016.pdf)).
+
+The TypeScript team had a choice between erasing types and enforcing types.
+They chose to erase types and run all code (typed or untyped) at the level
+ of JavaScript.
+(Some TypeScript [libraries](https://lorefnon.tech/2018/03/25/typescript-and-validations-at-runtime-boundaries/), however, can enforce some types.)
+
+> TypeScript is not the only erasure language, nor is it the first.
+> The oldest (I think) is [MACLISP](http://www.softwarepreservation.org/projects/LISP/maclisp_family/).
+> For an erasure manifesto, see [Pluggable Type Systems](http://bracha.org/pluggableTypesPosition.pdf).
+
+The Reticulated team faced a similar choice, and chose to enforce the top-level
+ shape of values in typed code ([POPL 2017](http://homes.sice.indiana.edu/mvitouse/papers/popl17.pdf)).
+It will be interesting to see if this guarantee helps developers maintain programs,
+ or if it is too shallow to be much use.
+The [Pyret](https://www.pyret.org/index.html) language has been successful with
+ a similar, shallow approach.
+
+> Note: the POPL 2017 paper advertises an "open-world soundness", but I do not
+> see how this idea is different from the older idea of soundness in a
+> multi-language system ([TOPLAS 2009](https://www.eecs.northwestern.edu/~robby/pubs/papers/toplas09-mf.pdf), [DLS 2006](https://www2.ccs.neu.edu/racket/pubs/dls06-tf.pdf)).
+
+The Java team faced a _different_ choice because the Java Virtual Machine is
+ a typed programming language.
+The casts around type-erased generics provide a minimal level of safety
+ --- without casts, use of a generic object might corrupt the state of a VM instance.
+Nevertheless, the implementation of generic-type erasure + cast insertion
+ is very similar to Reticulated's implementation of stronger guarantees for Python.
+
+Alternatively the Java team could have enforced generic types at run-time.
+Over the years there have been a few proposals to do so ([example 1](http://gafter.blogspot.com/2006/11/reified-generics-for-java.html),
```

Comment:
It would also be worthwhile to talk about C#, which does retain generic information at runtime with a type system that's otherwise very similar to Java's.

Reply:
Here's a PLDI paper from 2001: https://doi.org/10.1145/378795.378797

Non-paywalled link from MSR (unfortunately, I've had MSR break URLs on me in the past): https://www.microsoft.com/en-us/research/publication/design-and-implementation-of-generics-for-the-net-common-language-runtime/

I also found a blog post where someone explained the implementation details: https://mattwarren.org/2018/03/02/How-generics-were-added-to-.NET/